File: apps/blabee/spikes/m1/runtime-qualification/runtimes/swift/Coordinator.swift
import Dispatch
import Foundation
import CoreFoundation

#if canImport(Darwin)
import Darwin
#elseif canImport(Glibc)
import Glibc
#endif

let protocolVersion = 1
let buildVersion = "t005-qualification-v1"
let runtimeName = "swift"
let nominalWaitMilliseconds = 120_000

struct RuntimeFailure: Error, CustomStringConvertible {
    let description: String

    init(_ description: String) {
        self.description = description
    }
}

func journalPath(arguments: [String]) throws -> String {
    guard let argument = arguments.first(where: { $0.hasPrefix("--journal=") }) else {
        throw RuntimeFailure("--journal=<path> is required")
    }
    return String(argument.dropFirst("--journal=".count))
}

func writeJSON(_ object: [String: Any], to handle: FileHandle) throws {
    let data = try JSONSerialization.data(withJSONObject: object, options: [.sortedKeys])
    handle.write(data)
    handle.write(Data([0x0a]))
}

func log(_ event: String, fields: [String: Any] = [:]) {
    var record: [String: Any] = [
        "schema_version": "blabee.t005.runtime-log.v1",
        "event": event,
        "runtime": runtimeName,
    ]
    for (key, value) in fields {
        record[key] = value
    }
    try? writeJSON(record, to: FileHandle.standardError)
}

func respond(_ response: [String: Any]) {
    var envelope = response
    envelope["runtime"] = runtimeName
    envelope["protocol_version"] = protocolVersion
    do {
        try writeJSON(envelope, to: FileHandle.standardOutput)
    } catch {
        log("response_write_failed", fields: ["error": String(describing: error)])
    }
}

func positiveSequence(_ value: Any?) -> Int? {
    guard let number = value as? NSNumber else {
        return nil
    }
    guard CFGetTypeID(number) != CFBooleanGetTypeID() else {
        return nil
    }
    let doubleValue = number.doubleValue
    let intValue = number.intValue
    guard doubleValue == Double(intValue), intValue > 0 else {
        return nil
    }
    return intValue
}

func validateEvent(_ event: Any?, expectedSequence: Int) throws -> [String: Any] {
    guard let object = event as? [String: Any] else {
        throw RuntimeFailure("event must be an object")
    }
    guard let eventID = object["event_id"] as? String, !eventID.isEmpty else {
        throw RuntimeFailure("event_id must be a non-empty string")
    }
    guard let sequence = positiveSequence(object["event_sequence"]) else {
        throw RuntimeFailure("event_sequence must be a positive integer")
    }
    guard sequence == expectedSequence else {
        throw RuntimeFailure(
            "event_sequence \(sequence) does not match expected \(expectedSequence)"
        )
    }
    return object
}

final class DurableJournal {
    let path: String
    let handle: FileHandle
    private(set) var events: [[String: Any]] = []
    private(set) var partialTailBytes = 0

    init(path: String) throws {
        self.path = path
        let fileManager = FileManager.default
        let directory = URL(fileURLWithPath: path).deletingLastPathComponent().path
        try fileManager.createDirectory(
            atPath: directory,
            withIntermediateDirectories: true
        )
        if !fileManager.fileExists(atPath: path) {
            guard fileManager.createFile(atPath: path, contents: nil) else {
                throw RuntimeFailure("failed to create journal")
            }
        }

        self.handle = try FileHandle(forUpdating: URL(fileURLWithPath: path))
        let bytes = try Data(contentsOf: URL(fileURLWithPath: path))
        let lastNewline = bytes.lastIndex(of: 0x0a)
        let completeLength = lastNewline.map { bytes.distance(from: bytes.startIndex, to: $0) + 1 } ?? 0
        self.partialTailBytes = bytes.count - completeLength

        if partialTailBytes > 0 {
            try handle.truncate(atOffset: UInt64(completeLength))
            try handle.synchronize()
            log(
                "journal_partial_tail_truncated",
                fields: ["partial_tail_bytes": partialTailBytes]
            )
        }

        guard completeLength > 0 else {
            return
        }
        let completeData = bytes.prefix(completeLength)
        let lines = completeData.split(separator: 0x0a, omittingEmptySubsequences: false)
        for (index, line) in lines.dropLast().enumerated() {
            guard !line.isEmpty else {
                throw RuntimeFailure("journal line \(index + 1) is empty")
            }
            let parsed: Any
            do {
                parsed = try JSONSerialization.jsonObject(with: Data(line))
            } catch {
                throw RuntimeFailure(
                    "journal line \(index + 1) is invalid JSON: \(error)"
                )
            }
            events.append(try validateEvent(parsed, expectedSequence: events.count + 1))
        }
    }

    deinit {
        try? handle.close()
    }

    func append(_ rawEvent: Any?) throws -> [String: Any] {
        let event = try validateEvent(rawEvent, expectedSequence: events.count + 1)
        let data = try JSONSerialization.data(withJSONObject: event, options: [.sortedKeys])
        try handle.seekToEnd()
        try handle.write(contentsOf: data)
        try handle.write(contentsOf: Data([0x0a]))
        try handle.synchronize()
        events.append(event)
        log(
            "journal_append_durable",
            fields: [
                "event_id": event["event_id"] as! String,
                "event_sequence": event["event_sequence"] as! NSNumber,
            ]
        )
        return event
    }
}

func residentSetBytes() -> UInt64? {
    var usage = rusage()
    guard getrusage(RUSAGE_SELF, &usage) == 0 else {
        return nil
    }
    #if canImport(Darwin)
    return UInt64(usage.ru_maxrss)
    #else
    return UInt64(usage.ru_maxrss) * 1024
    #endif
}

func diagnostics(_ journal: DurableJournal) -> [String: Any] {
    let lastEvent = journal.events.last
    return [
        "ok": true,
        "build_version": buildVersion,
        "journal_event_count": journal.events.count,
        "last_event_id": lastEvent?["event_id"] ?? NSNull(),
        "last_event_sequence": lastEvent?["event_sequence"] ?? NSNull(),
        "replayed_event_count": journal.events.count,
        "partial_tail_truncated": journal.partialTailBytes > 0,
        "partial_tail_bytes": journal.partialTailBytes,
        "rss_bytes": residentSetBytes() ?? NSNull(),
        "rss_kind": "peak_resident_set_bytes",
        "pid": ProcessInfo.processInfo.processIdentifier,
    ]
}

func handleRequest(_ request: [String: Any], journal: DurableJournal) -> [String: Any] {
    guard let method = request["method"] as? String else {
        return ["ok": false, "error": "invalid_request"]
    }

    switch method {
    case "health":
        return ["ok": true]
    case "append":
        do {
            let event = try journal.append(request["event"])
            return [
                "ok": true,
                "durable": true,
                "event_id": event["event_id"] as! String,
                "event_sequence": event["event_sequence"] as! NSNumber,
                "journal_event_count": journal.events.count,
            ]
        } catch {
            return [
                "ok": false,
                "error": "append_rejected",
                "detail": String(describing: error),
            ]
        }
    case "diagnostics":
        return diagnostics(journal)
    case "update_info":
        return [
            "ok": true,
            "build_version": buildVersion,
            "update_strategy": "external_atomic_replacement_not_implemented",
        ]
    case "wait_probe":
        guard
            positiveSequence(request["nominal_wait_ms"]) == nominalWaitMilliseconds,
            let divisor = positiveSequence(request["scale_divisor"]),
            divisor <= nominalWaitMilliseconds
        else {
            return ["ok": false, "error": "invalid_wait_probe"]
        }
        let scaledWaitMilliseconds = Double(nominalWaitMilliseconds) / Double(divisor)
        let started = DispatchTime.now().uptimeNanoseconds
        Thread.sleep(forTimeInterval: scaledWaitMilliseconds / 1_000)
        let ended = DispatchTime.now().uptimeNanoseconds
        let elapsedMilliseconds = Double(ended - started) / 1_000_000
        log(
            "wait_probe_completed",
            fields: [
                "nominal_wait_ms": nominalWaitMilliseconds,
                "scale_divisor": divisor,
                "elapsed_monotonic_ms": elapsedMilliseconds,
            ]
        )
        return [
            "ok": true,
            "nominal_wait_ms": nominalWaitMilliseconds,
            "scale_divisor": divisor,
            "scaled_wait_ms": scaledWaitMilliseconds,
            "elapsed_monotonic_ms": elapsedMilliseconds,
            "automatic_selection": false,
        ]
    case "shutdown":
        return ["ok": true, "shutdown": true]
    default:
        return ["ok": false, "error": "unsupported_method"]
    }
}

do {
    let path = try journalPath(arguments: Array(CommandLine.arguments.dropFirst()))
    let journal = try DurableJournal(path: path)
    log(
        "runtime_started",
        fields: [
            "build_version": buildVersion,
            "journal_event_count": journal.events.count,
        ]
    )

    while let line = readLine() {
        guard
            let data = line.data(using: .utf8),
            let object = try? JSONSerialization.jsonObject(with: data),
            let request = object as? [String: Any]
        else {
            respond(["ok": false, "error": "invalid_json"])
            continue
        }

        let response = handleRequest(request, journal: journal)
        respond(response)
        if request["method"] as? String == "shutdown" {
            break
        }
    }
} catch {
    log("runtime_start_failed", fields: ["error": String(describing: error)])
    exit(2)
}

File: apps/blabee/spikes/m1/runtime-qualification/runtimes/node/coordinator.mjs
import {
  closeSync,
  fstatSync,
  fsyncSync,
  ftruncateSync,
  mkdirSync,
  openSync,
  readFileSync,
  writeSync,
} from "node:fs";
import { dirname, resolve } from "node:path";
import { createInterface } from "node:readline";

const PROTOCOL_VERSION = 1;
const BUILD_VERSION = "t005-qualification-v1";
const RUNTIME = "node";
const NOMINAL_WAIT_MS = 120_000;

function parseJournalPath(argv) {
  const argument = argv.find((value) => value.startsWith("--journal="));
  if (!argument) {
    throw new Error("--journal=<path> is required");
  }
  return resolve(argument.slice("--journal=".length));
}

function log(event, fields = {}) {
  process.stderr.write(
    `${JSON.stringify({
      schema_version: "blabee.t005.runtime-log.v1",
      event,
      runtime: RUNTIME,
      ...fields,
    })}\n`,
  );
}

function writeResponse(response) {
  process.stdout.write(
    `${JSON.stringify({
      ...response,
      runtime: RUNTIME,
      protocol_version: PROTOCOL_VERSION,
    })}\n`,
  );
}

function parseJournal(path) {
  mkdirSync(dirname(path), { recursive: true });
  const descriptor = openSync(path, "a+");
  const bytes = readFileSync(descriptor);
  const lastNewline = bytes.lastIndexOf(0x0a);
  const completeLength = lastNewline === -1 ? 0 : lastNewline + 1;
  const partialTailBytes = bytes.length - completeLength;

  if (partialTailBytes > 0) {
    ftruncateSync(descriptor, completeLength);
    fsyncSync(descriptor);
    log("journal_partial_tail_truncated", { partial_tail_bytes: partialTailBytes });
  }

  const events = [];
  if (completeLength > 0) {
    const lines = bytes.subarray(0, completeLength).toString("utf8").split("\n");
    for (let index = 0; index < lines.length - 1; index += 1) {
      const line = lines[index];
      if (line.length === 0) {
        throw new Error(`journal line ${index + 1} is empty`);
      }
      let event;
      try {
        event = JSON.parse(line);
      } catch (error) {
        throw new Error(`journal line ${index + 1} is invalid JSON: ${error.message}`);
      }
      validateEvent(event, events.length + 1);
      events.push(event);
    }
  }

  return {
    descriptor,
    events,
    partialTailBytes,
  };
}

function validateEvent(event, expectedSequence) {
  if (!event || typeof event !== "object" || Array.isArray(event)) {
    throw new Error("event must be an object");
  }
  if (typeof event.event_id !== "string" || event.event_id.length === 0) {
    throw new Error("event_id must be a non-empty string");
  }
  if (!Number.isSafeInteger(event.event_sequence) || event.event_sequence < 1) {
    throw new Error("event_sequence must be a positive safe integer");
  }
  if (event.event_sequence !== expectedSequence) {
    throw new Error(
      `event_sequence ${event.event_sequence} does not match expected ${expectedSequence}`,
    );
  }
}

function appendDurably(journal, event) {
  validateEvent(event, journal.events.length + 1);
  const bytes = Buffer.from(`${JSON.stringify(event)}\n`, "utf8");
  const offset = fstatSync(journal.descriptor).size;
  const written = writeSync(journal.descriptor, bytes, 0, bytes.length, offset);
  if (written !== bytes.length) {
    throw new Error(`short journal write: ${written}/${bytes.length}`);
  }
  fsyncSync(journal.descriptor);
  journal.events.push(event);
  log("journal_append_durable", {
    event_id: event.event_id,
    event_sequence: event.event_sequence,
  });
}

function diagnostics(journal) {
  const lastEvent = journal.events.at(-1) ?? null;
  return {
    ok: true,
    build_version: BUILD_VERSION,
    journal_event_count: journal.events.length,
    last_event_id: lastEvent?.event_id ?? null,
    last_event_sequence: lastEvent?.event_sequence ?? null,
    replayed_event_count: journal.events.length,
    partial_tail_truncated: journal.partialTailBytes > 0,
    partial_tail_bytes: journal.partialTailBytes,
    rss_bytes: process.memoryUsage().rss,
    rss_kind: "current_resident_set_bytes",
    pid: process.pid,
  };
}

async function handleRequest(request, journal) {
  if (!request || typeof request !== "object" || Array.isArray(request)) {
    return { ok: false, error: "invalid_request" };
  }

  switch (request.method) {
    case "health":
      return { ok: true };
    case "append": {
      try {
        appendDurably(journal, request.event);
      } catch (error) {
        return { ok: false, error: "append_rejected", detail: error.message };
      }
      return {
        ok: true,
        durable: true,
        event_id: request.event.event_id,
        event_sequence: request.event.event_sequence,
        journal_event_count: journal.events.length,
      };
    }
    case "diagnostics":
      return diagnostics(journal);
    case "update_info":
      return {
        ok: true,
        build_version: BUILD_VERSION,
        update_strategy: "external_atomic_replacement_not_implemented",
      };
    case "wait_probe": {
      const divisor = request.scale_divisor;
      if (
        request.nominal_wait_ms !== NOMINAL_WAIT_MS ||
        !Number.isInteger(divisor) ||
        divisor < 1 ||
        divisor > NOMINAL_WAIT_MS
      ) {
        return { ok: false, error: "invalid_wait_probe" };
      }
      const scaledWaitMs = NOMINAL_WAIT_MS / divisor;
      const started = process.hrtime.bigint();
      await new Promise((resolvePromise) => setTimeout(resolvePromise, scaledWaitMs));
      const elapsedMs = Number(process.hrtime.bigint() - started) / 1_000_000;
      log("wait_probe_completed", {
        nominal_wait_ms: NOMINAL_WAIT_MS,
        scale_divisor: divisor,
        elapsed_monotonic_ms: elapsedMs,
      });
      return {
        ok: true,
        nominal_wait_ms: NOMINAL_WAIT_MS,
        scale_divisor: divisor,
        scaled_wait_ms: scaledWaitMs,
        elapsed_monotonic_ms: elapsedMs,
        automatic_selection: false,
      };
    }
    case "shutdown":
      return { ok: true, shutdown: true };
    default:
      return { ok: false, error: "unsupported_method" };
  }
}

async function main() {
  const journalPath = parseJournalPath(process.argv.slice(2));
  const journal = parseJournal(journalPath);
  log("runtime_started", {
    build_version: BUILD_VERSION,
    journal_event_count: journal.events.length,
  });

  const input = createInterface({ input: process.stdin, crlfDelay: Infinity });
  try {
    for await (const line of input) {
      let request;
      try {
        request = JSON.parse(line);
      } catch {
        writeResponse({ ok: false, error: "invalid_json" });
        continue;
      }

      const response = await handleRequest(request, journal);
      writeResponse(response);
      if (request.method === "shutdown") {
        break;
      }
    }
  } finally {
    closeSync(journal.descriptor);
  }
}

main().catch((error) => {
  log("runtime_start_failed", { error: error.message });
  process.exitCode = 2;
});
